Instant-exit detection defeated by claude-wrapper pre-flight output

## Problem

`_is_instant_exit()` in `base.py` fails to detect sessions where the Claude CLI dies immediately (0s duration, SIGKILL) because the claude-wrapper's pre-flight log messages are captured by pipe-pane and exceed the `INSTANT_EXIT_MIN_OUTPUT_CHARS = 100` character threshold.

## Observed behavior

During a shepherd run for spheresemi/sphere#386 (`/shepherd 386 -m`), the judge phase failed after 3 validation retries:

```
[ERROR] Judge phase failed after 3 retry attempt(s): judge phase validation failed: 
  no judge comments detected; labels=[loom:reviewing, loom:review-requested]; 
  last output: '/Users/rwalters/GitHub/sphere/.loom/scripts/claude-wrapper.sh: line 424: 
  19130 Killed: 9               sleep 5'; session duration: 0s
```

**Key details:**
- Session duration: **0s** (Claude CLI died immediately with SIGKILL)
- The `loom:reviewing` label was applied (judge claimed the PR) but no review was produced
- This happened **3 times consistently** — not a transient issue
- The `Killed: 9` on `sleep 5` was the output monitor background process being killed when the wrapper died

## Root cause chain

1. `agent-spawn.sh` creates tmux session with pipe-pane log capture
2. `claude-wrapper.sh` starts and writes pre-flight messages to terminal:
   - `"[2026-02-16 21:22:21] [INFO] Attempt 1/5: Starting Claude CLI"`
   - `"[2026-02-16 21:22:21] [INFO] Pre-flight checks passed"` 
   - (etc.)
3. These messages are captured by pipe-pane into `.loom/logs/loom-judge-issue-386.log`
4. Claude CLI starts but is killed immediately (SIGKILL, 0s)
5. Wrapper exits, output monitor's `sleep 5` is killed
6. `agent-wait-bg.sh` returns exit 0 (session completed)
7. `_is_instant_exit()` checks: `len(non_header.strip()) < 100` → **False** (wrapper pre-flight output > 100 chars)
8. `run_worker_phase()` returns exit 0 (appears successful)
9. Judge validation fails → force-mode fallbacks fail → shepherd fails

## Why existing detection is insufficient

Issue #2381 noted that the `st_mtime - st_ctime` duration check in `_is_instant_exit()` is dead code, and that the output-size guard saves it. But the output-size guard only works when the **only** output is the log header (which starts with `#` and is excluded). When claude-wrapper writes pre-flight output to the terminal (captured by pipe-pane), this non-header text exceeds 100 chars, defeating the guard.

## Proposed fix

Option A: **Distinguish wrapper output from Claude CLI output** — have the wrapper write a sentinel marker (e.g., `# CLAUDE_CLI_START`) to the log when the actual `claude` command begins. `_is_instant_exit` would then only count output after this marker.

Option B: **Raise the threshold significantly** — increase `INSTANT_EXIT_MIN_OUTPUT_CHARS` to something like 500-1000 to account for wrapper boilerplate. Less precise but simpler.

Option C: **Check for agent-specific output markers** — look for patterns that indicate the Claude agent actually did work (e.g., tool calls, response text) rather than just counting raw characters.

## Environment

- macOS Darwin 25.2.0
- Claude Code (Claude Opus 4.6)
- Shepherd running inside Claude Code session, spawning judge via tmux

## Related issues

- #2135 — Original instant-exit detection (closed, but the fix doesn't cover this case)
- #2139 — Judge 0s duration (closed, same root cause variant)
- #2381 — `st_mtime - st_ctime` metric is broken (noted output-size guard as workaround)

--- a/defaults/scripts/claude-wrapper.sh
+++ b/defaults/scripts/claude-wrapper.sh
@@ -571,6 +571,10 @@ run_with_retry() {
         # Falls back to direct execution when no TTY is available (e.g., when spawned
         # from Claude Code's Bash tool rather than a tmux terminal).
         start_output_monitor "${temp_output}" "${monitor_pid_file}"
+        # Write sentinel marker so _is_instant_exit() in the shepherd can
+        # distinguish wrapper pre-flight output from actual Claude CLI output.
+        # The "# " prefix means it is also filtered as a header line.
+        echo "# CLAUDE_CLI_START" >&2
         set +e  # Temporarily disable errexit to capture exit code
         unset CLAUDECODE  # Prevent nested session guard from blocking subprocess
         # Export per-agent config dir if set (for session isolation)

--- a/loom-tools/src/loom_tools/shepherd/phases/base.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/base.py
@@ -23,8 +23,13 @@ _HEARTBEAT_POLL_INTERVAL = 5
 
 # Minimum characters of non-header content required for "meaningful output".
 # Sessions with less than this are treated as transient spawn failures
-# (e.g., Claude API error on startup).  See issues #2135, #2381.
+# (e.g., Claude API error on startup).  See issues #2135, #2381, #2401.
 INSTANT_EXIT_MIN_OUTPUT_CHARS = 100
+
+# Sentinel line written by claude-wrapper.sh just before invoking the Claude
+# CLI.  Output before this marker is wrapper pre-flight boilerplate and should
+# be excluded when measuring meaningful output.  See issue #2401.
+_CLI_START_SENTINEL = "# CLAUDE_CLI_START"
 
 # Maximum retries for instant-exit detection, with exponential backoff.
 INSTANT_EXIT_MAX_RETRIES = 3
@@ -291,6 +296,35 @@ def _print_heartbeat(action: str) -> None:
     print(f"\033[2m[{ts}] \u27f3 {action}\033[0m", file=sys.stderr)
 
 
+def _get_cli_output(stripped: str) -> str:
+    """Extract non-header output produced after the CLI start sentinel.
+
+    If the ``# CLAUDE_CLI_START`` sentinel is present, only lines after the
+    **last** occurrence are considered (the wrapper may emit multiple sentinels
+    when retrying).  Lines starting with ``# `` are always excluded as log
+    headers.
+
+    If no sentinel is found, falls back to the previous behaviour of
+    considering all non-header lines.
+
+    Args:
+        stripped: ANSI-stripped log file content.
+
+    Returns:
+        The meaningful (non-header, post-sentinel) output as a single string.
+    """
+    lines = stripped.splitlines()
+
+    # Find the last sentinel index.
+    sentinel_idx: int | None = None
+    for i, line in enumerate(lines):
+        if line.strip() == _CLI_START_SENTINEL:
+            sentinel_idx = i
+
+    start = (sentinel_idx + 1) if sentinel_idx is not None else 0
+    return "\n".join(line for line in lines[start:] if not line.startswith("# "))
+
+
 def _is_mcp_failure(log_path: Path) -> bool:
     """Check if a session log indicates an MCP server initialization failure.
 
@@ -326,16 +360,15 @@ def _is_mcp_failure(log_path: Path) -> bool:
         content = log_path.read_text()
         stripped = strip_ansi(content)
 
-        # If the session produced substantial output beyond headers,
-        # it was productive — MCP text is just status bar noise.
-        non_header = "\n".join(
-            line for line in stripped.splitlines() if not line.startswith("# ")
-        )
-        if len(non_header.strip()) >= MCP_FAILURE_MIN_OUTPUT_CHARS:
+        # If the session produced substantial CLI output beyond headers
+        # and wrapper pre-flight, it was productive — MCP text is just
+        # status bar noise.  See issues #2374, #2381, #2401.
+        cli_output = _get_cli_output(stripped)
+        if len(cli_output.strip()) >= MCP_FAILURE_MIN_OUTPUT_CHARS:
             return False
 
         for pattern in MCP_FAILURE_PATTERNS:
-            if re.search(pattern, stripped, re.IGNORECASE):
+            if re.search(pattern, cli_output, re.IGNORECASE):
                 return True
     except OSError:
         pass
@@ -369,12 +402,11 @@ def _is_instant_exit(log_path: Path) -> bool:
     try:
         content = log_path.read_text()
         stripped = strip_ansi(content)
-        # Exclude log header lines (e.g. "# Loom Agent Log", "# Session: ...")
-        # so that a log with only the header is correctly detected as instant exit.
-        non_header = "\n".join(
-            line for line in stripped.splitlines() if not line.startswith("# ")
-        )
-        return len(non_header.strip()) < INSTANT_EXIT_MIN_OUTPUT_CHARS
+        # Exclude log header lines and wrapper pre-flight output (everything
+        # before the last ``# CLAUDE_CLI_START`` sentinel) so that only
+        # actual Claude CLI output counts.  See issues #2135, #2381, #2401.
+        cli_output = _get_cli_output(stripped)
+        return len(cli_output.strip()) < INSTANT_EXIT_MIN_OUTPUT_CHARS
     except OSError:
         return False
 

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -9860,6 +9860,64 @@ class TestIsInstantExit:
         log.write_text("")
         assert _is_instant_exit(log) is True
 
+    def test_wrapper_preflight_with_sentinel_and_no_cli_output(
+        self, tmp_path: Path
+    ) -> None:
+        """Wrapper pre-flight output before sentinel should be ignored (issue #2401).
+
+        When the claude-wrapper writes pre-flight messages that exceed the
+        100-char threshold but the CLI itself produces no output, the session
+        should still be detected as an instant exit.
+        """
+        log = tmp_path / "session.log"
+        log.write_text(
+            "# Loom Agent Log\n"
+            "# Session: loom-judge-issue-42\n"
+            "[2026-02-16 21:22:21] [INFO] Claude wrapper starting\n"
+            "[2026-02-16 21:22:21] [INFO] Arguments: --resume\n"
+            "[2026-02-16 21:22:21] [INFO] Workspace: /path/to/repo\n"
+            "[2026-02-16 21:22:21] [INFO] Running pre-flight checks...\n"
+            "[2026-02-16 21:22:21] [INFO] Pre-flight checks passed\n"
+            "[2026-02-16 21:22:21] [INFO] Attempt 1/5: Starting Claude CLI\n"
+            "# CLAUDE_CLI_START\n"
+        )
+        assert _is_instant_exit(log) is True
+
+    def test_wrapper_preflight_with_sentinel_and_real_output(
+        self, tmp_path: Path
+    ) -> None:
+        """Real CLI output after sentinel should NOT be an instant exit."""
+        log = tmp_path / "session.log"
+        log.write_text(
+            "[2026-02-16 21:22:21] [INFO] Claude wrapper starting\n"
+            "[2026-02-16 21:22:21] [INFO] Pre-flight checks passed\n"
+            "[2026-02-16 21:22:21] [INFO] Attempt 1/5: Starting Claude CLI\n"
+            "# CLAUDE_CLI_START\n"
+            + "x" * (INSTANT_EXIT_MIN_OUTPUT_CHARS + 1)
+        )
+        assert _is_instant_exit(log) is False
+
+    def test_no_sentinel_backwards_compat(self, tmp_path: Path) -> None:
+        """Without sentinel, all non-header output is counted (backward compat)."""
+        log = tmp_path / "session.log"
+        # Pre-flight output exceeds threshold — without sentinel this is NOT
+        # detected as instant exit (the pre-2401 behaviour).
+        log.write_text(
+            "# Loom Agent Log\n"
+            + "[INFO] wrapper line\n" * 20
+        )
+        assert _is_instant_exit(log) is False
+
+    def test_multiple_sentinels_uses_last(self, tmp_path: Path) -> None:
+        """With multiple sentinels (retries), only output after the last counts."""
+        log = tmp_path / "session.log"
+        log.write_text(
+            "# CLAUDE_CLI_START\n"
+            + "x" * (INSTANT_EXIT_MIN_OUTPUT_CHARS + 1) + "\n"
+            "# CLAUDE_CLI_START\n"  # Second attempt — no output after this
+        )
+        assert _is_instant_exit(log) is True
+
 
 class TestRunWorkerPhaseInstantExit:
     """Test that run_worker_phase detects instant exits and returns code 6."""
@@ -10569,6 +10627,35 @@ class TestIsMcpFailure:
         log = tmp_path / "session.log"
         real_work = "x" * MCP_FAILURE_MIN_OUTPUT_CHARS
         log.write_text(f"# Header\n{real_work}\n1 MCP server failed\n")
+        assert _is_mcp_failure(log) is False
+
+    def test_wrapper_preflight_with_sentinel_and_mcp_failure(
+        self, tmp_path: Path
+    ) -> None:
+        """MCP failure after sentinel should be detected even with pre-flight output."""
+        log = tmp_path / "session.log"
+        log.write_text(
+            "[INFO] Claude wrapper starting\n"
+            "[INFO] Pre-flight checks passed\n"
+            "[INFO] Attempt 1/5: Starting Claude CLI\n"
+            "# CLAUDE_CLI_START\n"
+            "1 MCP server failed\n"
+        )
+        assert _is_mcp_failure(log) is True
+
+    def test_wrapper_preflight_with_sentinel_and_productive_session(
+        self, tmp_path: Path
+    ) -> None:
+        """Productive CLI output after sentinel should NOT be flagged as MCP failure."""
+        log = tmp_path / "session.log"
+        real_work = "Working on issue...\n" * 40
+        log.write_text(
+            "[INFO] Claude wrapper starting\n"
+            "# CLAUDE_CLI_START\n"
+            "1 MCP server failed\n"
+            f"{real_work}"
+        )
+        assert len(real_work) >= MCP_FAILURE_MIN_OUTPUT_CHARS
         assert _is_mcp_failure(log) is False
 
 

